src/reducers/url-state.js

```diff
@@ -186,16 +187,59 @@ const networkSearchString: Reducer<string> = (state = '', action) => {
   }
 };
 
-const transforms: Reducer<TransformStacksPerThread> = (state = {}, action) => {
+const transforms: Reducer<TransformStacksPerThread> = (
+  state: TransformStacksPerThread = {},
+  action
+) => {
   switch (action.type) {
     case 'PROFILE_LOADED':
       return action.transformStacks || state;
     case 'ADD_TRANSFORM_TO_STACK': {
       const { threadsKey, transform } = action;
-      const transforms = state[threadsKey] || [];
-      return Object.assign({}, state, {
-        [threadsKey]: [...transforms, transform],
-      });
+      const oldTransforms: TransformStack = state[threadsKey] || [];
+      let newTransforms: TransformStack | null = null;
+
+      // Multiple "merge-function" transforms will combine into a "merge-function-set"
+      // transform, as the operation is commutative. This makes the transforms more
+      // performant when mashing the shortcut and radically adding many of them.
+      if (oldTransforms.length > 0 && transform.type === 'merge-function') {
+        const lastTransform = oldTransforms[oldTransforms.length - 1];
+        switch (lastTransform.type) {
+          case 'merge-function':
+            {
+              newTransforms = [...oldTransforms];
+              newTransforms.pop();
+              newTransforms.push({
+                type: 'merge-function-set',
+                funcIndexes: new Set([
+                  lastTransform.funcIndex,
+                  transform.funcIndex,
+                ]),
+              });
+            }
+            break;
```

Comment:
I wonder if instead of adding a new transform, we could change the existing mergeFunction to have this functionality. Indeed now `mergeFunction` looks a special case of `mergeFunctionSet`. And this "conversion" happening here wouldn't ben ecessary anymore.
Of course this probably means we'd need an url upgrader. What do you think?

Reply:
I guess the approach of using the merge-function-set means that the first merge function doesn't get hit with the small performance regression.

Reply:
I'm confused, what is the small performance regression you're talking about? Is that the use of a Set with only one element? This sounds minimal.

Because this would need an url upgrader, I guess I'm fine with landing this in this form though, and look at reducing the code complexity later.